### PR TITLE
fix: throw on retry when payload is consume by downstream

### LINKED ISSUE #3356: The RetryHandler receives a duplicate body when the server does not support Range requests.
```
import { createServer } from "node:http"
import { once } from "node:events"
import { RetryAgent, Agent, fetch } from "undici";
import asserts from "node:assert";

let retry = true;
const server = createServer((req, res) => {
  res.writeHead(200, { "Content-Type": "text/plain" });
	if (retry) {
		retry = false;
		res.flushHeaders();
		res.write('H');
		setTimeout(() => {
			res.end('ello World');
		}, 1000);
	} else {
		res.end('Hello World');
	}
})

server.listen();
await once(server, 'listening')

const agent = new RetryAgent(new Agent({bodyTimeout: 500}), {
	errorCodes: ['UND_ERR_BODY_TIMEOUT'],
});
const { port } = server.address()
const res = await fetch(`http://localhost:${port}`, {
    dispatcher: agent,
});
asserts.equal(await res.text(), 'Hello World');

server.close();
```

output:
```
AssertionError [ERR_ASSERTION]: 'HHello World' == 'Hello World'
```

### PR DESCRIPTION
**Pull request by climba03003**
Close #3365 Fixes #3356

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: lib/handler/retry-handler.js [hunk 1]
```diff
@@ -296,7 +310,10 @@ class RetryHandler {
   onData (chunk) {
     this.start += chunk.length
 
-    return this.handler.onData(chunk)
+    const written = this.handler.onData(chunk)
```
Comment (climba03003) [inline lib/handler/retry-handler.js:313]: I use the return of `onData` to indicate the data is written to downstream.
When it return `false` from the first write, we still have chance to retry.

Comment (ronag) [reply lib/handler/retry-handler.js:313]: You can't do that. It only indicated back pressure. Not whether the data was sent or not.

Comment (climba03003) [reply lib/handler/retry-handler.js:313]: Doesn't `false` means the data didn't push to the buffer?
If yes, then it is what we need for this case.

Comment (climba03003) [reply lib/handler/retry-handler.js]: Nevermind, I found the answer. It is no way back.

File: lib/handler/retry-handler.js [hunk 2]
```diff
@@ -193,6 +194,19 @@ class RetryHandler {
       this.resume = null
 
       if (statusCode !== 206) {
+        // Abort when status code is not partial content
+        // and payload is already consumed because downstream
+        // will concatenate response from two request wrongly.
```
Comment (mcollina) [inline lib/handler/retry-handler.js:199]: Can you please expand the comment, explaining that anything that is not a 206 implies that a range request cannot be made?

File: lib/handler/retry-handler.js [hunk 3]
```diff
@@ -294,6 +308,7 @@ class RetryHandler {
   }
 
   onData (chunk) {
+    this.consumed = true
```
Comment (mcollina) [inline lib/handler/retry-handler.js:311]: I don't think we need a new prop, we could just use .start

File: lib/handler/retry-handler.js [hunk 4]
```diff
@@ -193,6 +194,19 @@ class RetryHandler {
       this.resume = null
 
       if (statusCode !== 206) {
+        // Abort when status code is not partial content
+        // and payload is already consumed because downstream
+        // will concatenate response from two request wrongly.
+        if (this.consumed) {
+          this.abort(
+            new RequestRetryError('unsupported retry when payload is consumed', statusCode, {
```
Comment (mcollina) [inline lib/handler/retry-handler.js:202]: The error message is not very actionable. I would rephrase to something like
```suggestion
            new RequestRetryError('server does not support the range header and the payload was partially consumed', statusCode, {
```

File: lib/handler/retry-handler.js [hunk 5]
```diff
@@ -193,6 +194,19 @@ class RetryHandler {
       this.resume = null
 
       if (statusCode !== 206) {
```
Comment (ronag) [inline lib/handler/retry-handler.js:196]:
```suggestion
      if (statusCode !== 206 && (this.start > 0 || statusCode !== 200) {
```

File: lib/handler/retry-handler.js [hunk 6]
```diff
@@ -192,8 +192,18 @@ class RetryHandler {
...
+      // should not be retry because it would result in downstream
+      // wrongly concatanete multiple responses.
+      if (statusCode !== 206 && (this.start > 0 || statusCode !== 200)) {
+        this.abort(
+          new RequestRetryError('server does not support the range header and the payload was partially consumed', statusCode, {
+            headers,
+            data: { count: this.retryCount }
+          })
+        )
+        return false
```
Comment (ronag) [inline lib/handler/retry-handler.js:206]:
```suggestion
        throw new RequestRetryError('server does not support the range header and the payload was partially consumed', statusCode, {
            headers,
            data: { count: this.retryCount }
          })
        )
```

**Review by ronag — APPROVED** [review]

File: lib/handler/retry-handler.js (hunk 6 above)
Comment (climba03003) [reply lib/handler/retry-handler.js:206]: Are you sure it should throw instead of similar to the others error?

Comment (ronag) [reply lib/handler/retry-handler.js:206]: I think all errors should throw

Comment (ronag) [reply lib/handler/retry-handler.js:206]: it's a nit, you can leave it if you want

**Review by mcollina — APPROVED** [review]
lgtm
